gh-82054: allow test runner to split test_asyncio to execute in parallel by sharding.

zitterbewegung opened the pull request:
## Summary:

This runs test_asyncio sub-tests in parallel using sharding from Cinder. This suite is typically the longest-pole in runs because it is a test package with a lot of further sub-tests otherwise run serially. By breaking out the sub-tests as independent modules we can run a lot more in parallel.

After porting we can see the direct impact on a multicore system.
* Without this change:
  * Running make test is 5 min 26 seconds
* With this change:
  * Running make test takes 3 min 39 seconds

The drawbacks are that this implementation is hacky and due to the sorting of the tests it obscures when the asyncio tests occur and involves changing CPython test infrastructure but, the wall time saved it is worth it, especially in low-core count CI runs as it pulls a long tail. The win for productivity and reserved CI resource usage is significant.

Future tests that deserve to be refactored into split up suites to benefit from are `test_concurrent_futures` and the way the `_test_multiprocessing` suite gets run for all start methods.  As exposed by passing the `-o` flag to `python -m test` to get a list of the 10 longest running tests.

<!-- gh-issue-number: gh-82054 -->
* Issue: gh-82054
<!-- /gh-issue-number -->

zitterbewegung committed:
This allows the test runner to execute test_multiprocessing and test_asyncio in parallel.

By sharding the individual tests in test_multiprocessing and test_asyncio we are able to reduce considerably the time it takes to run the whole test suite.

carljm committed:
revert whitespace changes

carljm committed:
multiprocessing, not compiler

gpshead commented on Lib/test/libregrtest/runtest.py:
```diff
@@ -158,16 +167,24 @@ def findtestdir(path=None):
 
 
-def findtests(testdir=None, stdtests=STDTESTS, nottests=NOTTESTS):
+def findtests(testdir=None, stdtests=STDTESTS, nottests=NOTTESTS, splittestdirs=SPLITTESTDIRS, base_mod=""):
```
spell that split_test_dirs= and lets make the new args keyword only.

zitterbewegung replied:
Done

gpshead commented on Lib/test/libregrtest/runtest.py:
```diff
@@ -143,6 +143,15 @@ def __str__(self) -> str:
+
+SPLITTESTDIRS = {
+    "test_asyncio",
+    "test_multiprocessing",
```
there isn't a test_multiprocessing package to this line doesn't do anything.

zitterbewegung replied:
Okay ill remove.

gpshead reviewed (commented):
I think we've still got a while before this meaningfully reduces most CI system (presumed 2-4 thread) or developer (presumed 8-32 thread) whole test suite latency.  With this PR on a fast system:

```
10 slowest tests:
- test_concurrent_futures: 2 min 14 sec
- test_multiprocessing_spawn: 1 min 25 sec
- test_multiprocessing_forkserver: 1 min 9 sec
- test_signal: 1 min 7 sec
- test_tools: 1 min 3 sec
- test_multiprocessing_fork: 1 min 3 sec
- test_socket: 37.6 sec
- test_imaplib: 30.6 sec
- test_io: 29.7 sec
- test.test_asyncio.test_tasks: 28.8 sec
```

Until `Lib/test_concurrent_futures.py` is broken up into a package of parallelizable items we've still got a pretty long tail.

The same really should be done with `Lib/test/_test_multiprocessing.py` as well, but it's a more complicated beast as it is a single test "template" that gets invoked with three different start method settings configured by the `Lib/test/test_multiprocessing_{fork,forkserver,spawn}.py` files.

Regardless this PR is still a win even with our longest test, test_asyncio, which would otherwise be 3 minutes above, split up.  That reduces the total run time latency 10% on a CI-like `-j 4` run.

zitterbewegung committed:
This removes an extraneous test package that doesn't exist and also code cleanup.

For the new test arguments they were made keyword only and also test_multiprocessing was removed from the SPLITTESTDIRS. The argument splittestdirs was renamed to split_test_dirs for findtests.

gpshead committed:
simplify logic: no len()s, tight f-expression.
